File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/no-alias-methods.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-alias-methods', {} as never, {
  valid: [
    { code: 'expect(a).toHaveBeenCalled()' },
    { code: 'expect(a).toHaveBeenCalledTimes()' },
    { code: 'expect(a).toHaveBeenCalledWith()' },
    { code: 'expect(a).toHaveBeenLastCalledWith()' },
    { code: 'expect(a).toHaveBeenNthCalledWith()' },
    { code: 'expect(a).toHaveReturned()' },
    { code: 'expect(a).toHaveReturnedTimes()' },
    { code: 'expect(a).toHaveReturnedWith()' },
    { code: 'expect(a).toHaveLastReturnedWith()' },
    { code: 'expect(a).toHaveNthReturnedWith()' },
    { code: 'expect(a).toThrow()' },
    { code: 'expect(a).rejects;' },
    { code: 'expect(a);' },
  ],

  invalid: [
    {
      code: 'expect(a).toBeCalled()',
      output: 'expect(a).toHaveBeenCalled()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'toBeCalled',
            canonical: 'toHaveBeenCalled',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).toBeCalledTimes()',
      output: 'expect(a).toHaveBeenCalledTimes()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'toBeCalledTimes',
            canonical: 'toHaveBeenCalledTimes',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).toBeCalledWith()',
      output: 'expect(a).toHaveBeenCalledWith()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'toBeCalledWith',
            canonical: 'toHaveBeenCalledWith',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).lastCalledWith()',
      output: 'expect(a).toHaveBeenLastCalledWith()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'lastCalledWith',
            canonical: 'toHaveBeenLastCalledWith',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).nthCalledWith()',
      output: 'expect(a).toHaveBeenNthCalledWith()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'nthCalledWith',
            canonical: 'toHaveBeenNthCalledWith',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).toReturn()',
      output: 'expect(a).toHaveReturned()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'toReturn',
            canonical: 'toHaveReturned',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).toReturnTimes()',
      output: 'expect(a).toHaveReturnedTimes()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'toReturnTimes',
            canonical: 'toHaveReturnedTimes',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).toReturnWith()',
      output: 'expect(a).toHaveReturnedWith()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'toReturnWith',
            canonical: 'toHaveReturnedWith',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).lastReturnedWith()',
      output: 'expect(a).toHaveLastReturnedWith()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'lastReturnedWith',
            canonical: 'toHaveLastReturnedWith',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).nthReturnedWith()',
      output: 'expect(a).toHaveNthReturnedWith()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'nthReturnedWith',
            canonical: 'toHaveNthReturnedWith',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).toThrowError()',
      output: 'expect(a).toThrow()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'toThrowError',
            canonical: 'toThrow',
          },
          column: 11,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).resolves.toThrowError()',
      output: 'expect(a).resolves.toThrow()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'toThrowError',
            canonical: 'toThrow',
          },
          column: 20,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).rejects.toThrowError()',
      output: 'expect(a).rejects.toThrow()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'toThrowError',
            canonical: 'toThrow',
          },
          column: 19,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).not.toThrowError()',
      output: 'expect(a).not.toThrow()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'toThrowError',
            canonical: 'toThrow',
          },
          column: 15,
          line: 1,
        },
      ],
    },
    {
      code: 'expect(a).not["toThrowError"]()',
      output: 'expect(a).not["toThrow"]()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: {
            alias: 'toThrowError',
            canonical: 'toThrow',
          },
          column: 15,
          line: 1,
        },
      ],
    },
    // The fix range must skip the matcher's leading trivia, otherwise the
    // newline, comment or opening quote in front of the name is overwritten.
    {
      code: 'expect(a).\n  toBeCalled()',
      output: 'expect(a).\n  toHaveBeenCalled()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: { alias: 'toBeCalled', canonical: 'toHaveBeenCalled' },
          column: 3,
          line: 2,
        },
      ],
    },
    {
      code: 'expect(a) . /* c */ toBeCalled()',
      output: 'expect(a) . /* c */ toHaveBeenCalled()',
      errors: [
        {
          messageId: 'replaceAlias',
          data: { alias: 'toBeCalled', canonical: 'toHaveBeenCalled' },
          column: 21,
          line: 1,
        },
      ],
    },
    {
      // `toBeCalled` is a variable here, so the alias is reported but not
      // fixed; rewriting it would point the computed key at an identifier that
      // does not exist.
      code: "const toBeCalled = 'toBeCalled';\nexpect(a)[toBeCalled]();",
      errors: [
        {
          messageId: 'replaceAlias',
          data: { alias: 'toBeCalled', canonical: 'toHaveBeenCalled' },
          column: 11,
          line: 2,
        },
      ],
    },
    {
      // Previously emitted `expect(a)[/toHaveBeenCalled']()`, which is not
      // parseable JavaScript.
      code: "expect(a)[/* c */ 'toBeCalled']()",
      output: "expect(a)[/* c */ 'toHaveBeenCalled']()",
      errors: [
        {
          messageId: 'replaceAlias',
          data: { alias: 'toBeCalled', canonical: 'toHaveBeenCalled' },
          column: 19,
          line: 1,
        },
      ],
    },
  ],
});

File: internal/plugins/jest/rules/no_alias_methods/no_alias_methods_test.go
package no_alias_methods_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/no_alias_methods"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoAliasMethodsRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&no_alias_methods.NoAliasMethodsRule,
		[]rule_tester.ValidTestCase{
			{Code: "expect(a).toHaveBeenCalled()"},
			{Code: "expect(a).toHaveBeenCalledTimes()"},
			{Code: "expect(a).toHaveBeenCalledWith()"},
			{Code: "expect(a).toHaveBeenLastCalledWith()"},
			{Code: "expect(a).toHaveBeenNthCalledWith()"},
			{Code: "expect(a).toHaveReturned()"},
			{Code: "expect(a).toHaveReturnedTimes()"},
			{Code: "expect(a).toHaveReturnedWith()"},
			{Code: "expect(a).toHaveLastReturnedWith()"},
			{Code: "expect(a).toHaveNthReturnedWith()"},
			{Code: "expect(a).toThrow()"},
			{Code: "expect(a).rejects;"},
			{Code: "expect(a);"},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code:   "expect(a).toBeCalled()",
				Output: []string{"expect(a).toHaveBeenCalled()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 11, Line: 1},
				},
			},
			{
				Code:   "expect(a).toBeCalledTimes()",
				Output: []string{"expect(a).toHaveBeenCalledTimes()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 11, Line: 1},
				},
			},
			{
				Code:   "expect(a).toBeCalledWith()",
				Output: []string{"expect(a).toHaveBeenCalledWith()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 11, Line: 1},
				},
			},
			{
				Code:   "expect(a).lastCalledWith()",
				Output: []string{"expect(a).toHaveBeenLastCalledWith()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 11, Line: 1},
				},
			},
			{
				Code:   "expect(a).nthCalledWith()",
				Output: []string{"expect(a).toHaveBeenNthCalledWith()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 11, Line: 1},
				},
			},
			{
				Code:   "expect(a).toReturn()",
				Output: []string{"expect(a).toHaveReturned()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 11, Line: 1},
				},
			},
			{
				Code:   "expect(a).toReturnTimes()",
				Output: []string{"expect(a).toHaveReturnedTimes()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 11, Line: 1},
				},
			},
			{
				Code:   "expect(a).toReturnWith()",
				Output: []string{"expect(a).toHaveReturnedWith()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 11, Line: 1},
				},
			},
			{
				Code:   "expect(a).lastReturnedWith()",
				Output: []string{"expect(a).toHaveLastReturnedWith()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 11, Line: 1},
				},
			},
			{
				Code:   "expect(a).nthReturnedWith()",
				Output: []string{"expect(a).toHaveNthReturnedWith()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 11, Line: 1},
				},
			},
			{
				Code:   "expect(a).toThrowError()",
				Output: []string{"expect(a).toThrow()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 11, Line: 1},
				},
			},
			{
				Code:   "expect(a).resolves.toThrowError()",
				Output: []string{"expect(a).resolves.toThrow()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 20, Line: 1},
				},
			},
			{
				Code:   "expect(a).rejects.toThrowError()",
				Output: []string{"expect(a).rejects.toThrow()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 19, Line: 1},
				},
			},
			{
				Code:   "expect(a).not.toThrowError()",
				Output: []string{"expect(a).not.toThrow()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 15, Line: 1},
				},
			},
			{
				Code:   `expect(a).not["toThrowError"]()`,
				Output: []string{`expect(a).not["toThrow"]()`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 15, Line: 1},
				},
			},
			// Leading-trivia cases. The matcher's Pos() sits at the end of the
			// preceding token, so a fix range built from it swallows whatever
			// whitespace or comment separates the dot from the matcher name.
			{
				Code:   "expect(a).\n  toBeCalled()",
				Output: []string{"expect(a).\n  toHaveBeenCalled()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 3, Line: 2},
				},
			},
			{
				// Control: the newline belongs to the dot rather than to the
				// matcher, so this shape was already correct.
				Code:   "expect(a)\n  .toBeCalled()",
				Output: []string{"expect(a)\n  .toHaveBeenCalled()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 4, Line: 2},
				},
			},
			{
				Code:   "expect(a) . /* c */ toBeCalled()",
				Output: []string{"expect(a) . /* c */ toHaveBeenCalled()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 21, Line: 1},
				},
			},
			{
				Code:   "expect(a). // c\n  toBeCalled()",
				Output: []string{"expect(a). // c\n  toHaveBeenCalled()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 3, Line: 2},
				},
			},
			{
				// The hand-written `Pos()+1 / End()-1` offsets used to land
				// inside the comment here, stripping `*` instead of the opening
				// quote and emitting `expect(a)[/toHaveBeenCalled']()`, which
				// does not parse.
				Code:   "expect(a)[/* c */ 'toBeCalled']()",
				Output: []string{"expect(a)[/* c */ 'toHaveBeenCalled']()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 19, Line: 1},
				},
			},
			{
				Code:   "expect(a)[\n  'toBeCalled'\n]()",
				Output: []string{"expect(a)[\n  'toHaveBeenCalled'\n]()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 3, Line: 2},
				},
			},
			{
				Code:   "expect(a)[`toBeCalled`]()",
				Output: []string{"expect(a)[`toHaveBeenCalled`]()"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 11, Line: 1},
				},
			},
			{
				// Multiple aliases in one chain are each fixed in place.
				Code:   "expect(a).\n  resolves.\n  toBeCalledTimes(1)",
				Output: []string{"expect(a).\n  resolves.\n  toHaveBeenCalledTimes(1)"},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "replaceAlias", Column: 3, Line: 3},
				},
			},
		},
	)
}
